indimail-2.5/vchkpass.8
.TH vchkpass 8
.SH NAME
.PP
vchkpass \- checkpassword program for IndiMail authentication
.PP
systpass \- checkpassword program for System users

.SH SYNOPSYS
.PP
\fBvchkpass\fR \fBalternate_checkpassword_program(s)\fR \fB/bin/false\fR
.PP
\fBsystpass\fR \fBalternate_checkpassword_program(s)\fR \fB/bin/false\fR

.SH DESCRIPTION
\fBvchkpass\fR and \fBsystpass\fR supports authentication for ESMTP AUTH option in \fBqmail-smtpd(8)\fR,
allowing the LOGIN, PLAIN, and CRAM-MD5, CRAM-SHA1, CRAM-RIPEMD, DIGEST-MD5 AUTH types.
\fBvchkpass\fR looks up userid in IndiMail's Virtual Database while \fBsystpass\fR looks
up userids in the systems authentication database (/etc/passwd or /etc/shadow).
.PP
As reflected in the \fBqmail-smtpd(8)\fR man page, \fBqmail-smtpd(8)\fR must be invoked with
three arguments: hostname, path of the vchkpass/systpass program, and subprogram. If these
arguments are missing, \fBqmail-smtpd(8)\fR will not advertise availability of AUTH.
.PP
\fBvchkpass\fR/\fBsystpass\fR reads descriptor 3 through end of file and then closes
descriptor 3. There must be at most 512 bytes of data before end of file. If it invokes an
\fBalternate_checkpassword_program\fR, it opens a pipe duping the read descriptor as 3 for
the \fBalternate_checkpassword_program\fR. The write end of the pipe is duped to descriptor 4
and the data which was earlier read on descriptor 3 is written back on descriptor 4.
The alternate program is invoked when vchkpass doesn't support authentication for the
supplied login name.
.PP
For LOGIN, PLAIN AUTH types, the information supplied on descriptor 3 is a login name
terminated by \\0 and password terminated by \\0. \fBvchkpass\fR encrypts the password using
cyrpt(3) routine and compares it against the value present in the passwd database. The encrypted
password can be in DES, MD5, SHA-256 or SHA-512 hash.
.PP
For CRAM-MD5, CRAM-SHA1, CRAM-RIPEMD, DIGEST-MD5 AUTH type, the information supplied on
descriptor 3 is a login name terminated by \\0, a cram-md5 challenge terminated by \\0, and
a cram-md5 response terminated by \\0.  \fBvchkpass\fR/\fBsystpass\fR encrypts the
challenge with keyed MD5 using passwords from passwd database. It's compared with
response (3rd parameter) and if they are the same then vchkpass/systpass uses execvp to
run subprogram with the given arguments. If they differ then it returns -1.
.PP
If challenge and response differ, \fBvchkpass\fR/\fBsystpass\fR exits 1. If \fBvchkpass\fR/\fBsystpass\fR
is misused, it may instead exit 2. If there is a temporary problem checking the password,
\fBvchkpass\fR/\fBsystpass\fR exits 111.
.PP
hostname is simply used to form the CRAM-MD5, CRAM-SHA1, CRAM-RIPEMD, DIGEST-MD5 challenge.
\fBqmail-smtpd\fR invokes \fBvchkpass\fR, feeding it the username and password, in the case
of LOGIN or PLAIN, or the username, challenge, and response, in the case of CRAM-MD5,
CRAM-SHA1, CRAM-RIPEMD, DIGEST-MD5.  If the user checks out, \fBvchkpass\fR/\fBsystpass\fR
exits with a status of 0 for the user to be authenticated. Otherwise, subprogram should
exit with a non-zero status, or call an alternate checkpassword program. The last
subprogram can usually be /usr/bin/false (or /bin/false, depending on your flavor of OS).

.PP
\fBvchkpass\fR does not authenticate if user is inactive. A user is active if one logs in using
IMAP or POP3 protocol. If \fBALLOW_INACTIVE\fR is defined, \fBvchkpass\fR will proceed regardless
of the user's active or inactive status.

If the environment variable \fBDOMAIN_LIMITS\fR is defined, \fBvchkpass\fR validates domain
expiry and password expiry set by \fBvmoduser(1)\fR program.

\fBvchkpass\fR can execute a script defined by environment variable \fBPOSTAUTH\fR
.PP
MUAs and AUTH types tested with this software are:

.IP \[bu] 2
Netscape Communicator 4.76 - LOGIN & PLAIN
.IP \[bu]
Microsoft Outlook 2000 - LOGIN
.IP \[bu]
Microsoft Outlook Express 5 - LOGIN
.IP \[bu]
Eudora 5.0.2 – CRAM-MD5
.IP \[bu]
sylpheed 0.8.2 – LOGIN, PLAIN and CRAM-MD5
.IP \[bu]
evolution - LOGIN, PLAIN
.IP \[bu]
Thunderbird - LOGIN, PLAIN

.SH OPTIONS
\fBalternate_checkpassword_program\fR
Path to an alternate checkpassword utility which supports additional authentication
advertised by qmail-smtpd.

.SH RETURN VALUE
0 if authenticated, 1 on authentication failure or system problems.

.SH "SEE ALSO"
qmail-smtpd(8),
qmail-poppass(8),
sys-checkpwd(8),
ldap-checkpwd(8),
pam-checkpwd(8),
nssd(8),
pam-multi(8)


indimail-1.6.5/vchkpass.c
/*
 * $Log: vchkpass.c,v $
 * Revision 2.33  2009-11-15 14:40:32+05:30  Cprogrammer
 * use vauthOpen_user() if QUERY_CACHE is not set for an extended domain
 *
 * Revision 2.32  2009-11-15 12:26:53+05:30  Cprogrammer
 * reorganized code for better readability
 *
 * Revision 2.31  2009-09-23 15:00:24+05:30  Cprogrammer
 * change for new runcmmd
 *
 * Revision 2.30  2009-04-17 21:00:09+05:30  Cprogrammer
 * log database error
 *
 * Revision 2.29  2009-04-17 14:31:28+05:30  Cprogrammer
 * BUG - Security bug with authentication of non-existent users
 *
 * Revision 2.28  2008-12-19 11:53:39+05:30  Cprogrammer
 * exit with return status of POSTAUTH command
 *
 * Revision 2.27  2008-12-18 11:55:26+05:30  Cprogrammer
 * changed PASSWD_CACHE to query cache
 *
 * Revision 2.26  2008-11-06 15:06:06+05:30  Cprogrammer
 * changed PASSWD_CACHE to QUERY_CACHE
 *
 * Revision 2.25  2008-08-28 21:40:59+05:30  Cprogrammer
 * removed is_open = 0
 *
 * Revision 2.24  2008-08-24 17:44:20+05:30  Cprogrammer
 * removed redundant code
 *
 * Revision 2.23  2008-07-13 19:49:35+05:30  Cprogrammer
 * use ERESTART only if available
 *
 * Revision 2.22  2008-05-28 15:26:11+05:30  Cprogrammer
 * removed ldap module
 *
 * Revision 2.21  2005-12-29 22:51:56+05:30  Cprogrammer
 * use getEnvConfigStr to set variables from environment variables
 *
 * Revision 2.20  2005-12-21 09:52:15+05:30  Cprogrammer
 * make gcc 4 happy
 *
 * Revision 2.19  2005-07-07 09:00:14+05:30  Cprogrammer
 * allow inactive users to send mails if ALLOW_INACTIVE env variable is defined
 * allow users created only in ldap to send mails if ALLOW_INACTIVE is defined
 * execute script defined by POSTAUTH env variable after successful auth
 *
 * Revision 2.18  2005-01-22 00:42:30+05:30  Cprogrammer
 * case for NO_RELAY bit in pw_gid fixed
 *
 * Revision 2.17  2004-10-27 14:49:36+05:30  Cprogrammer
 * close mysql before exit
 *
 * Revision 2.16  2004-10-22 20:51:36+05:30  Cprogrammer
 * BUG - typo, changed USE_LDAP to USE_LDAP_PASSWD
 *
 * Revision 2.15  2004-07-12 22:48:40+05:30  Cprogrammer
 * made PASSWD_CACHE configurable through env variable
 *
 * Revision 2.14  2004-05-17 00:52:48+05:30  Cprogrammer
 * missing conditional compilation added
 *
 * Revision 2.13  2004-01-05 14:12:01+05:30  Cprogrammer
 * added ldap authentication code
 *
 * Revision 2.12  2002-12-11 10:29:09+05:30  Cprogrammer
 * added ERESTART check for errno
 *
 * Revision 2.11  2002-10-20 22:18:10+05:30  Cprogrammer
 * correction for compilation warning on solaris
 *
 * Revision 2.10  2002-09-01 20:54:01+05:30  Cprogrammer
 * added option to call next program on command line passing data on file descriptor 3
 *
 * Revision 2.9  2002-08-31 16:00:48+05:30  Cprogrammer
 * added more informative messages for system errors
 *
 * Revision 2.8  2002-08-31 14:27:43+05:30  Cprogrammer
 * corrected logic when passwd fails and next auth program needs to be exec'ed
 *
 * Revision 2.7  2002-08-30 23:29:04+05:30  Cprogrammer
 * incorrect size of buffers corrected
 *
 * Revision 2.6  2002-07-05 03:54:37+05:30  Cprogrammer
 * used sizeof operator for cross platform compatibility
 *
 * Revision 2.5  2002-05-12 01:20:35+05:30  Cprogrammer
 * added conditional compilation of debug message
 *
 * Revision 2.4  2002-05-11 00:19:18+05:30  Cprogrammer
 * Implemented CRAM-MD5 authentication
 *
 * Revision 2.3  2002-05-10 10:08:37+05:30  Cprogrammer
 * invoke alternate checkpassword program only if authentication fails
 *
 * Revision 2.2  2002-04-15 21:08:16+05:30  Cprogrammer
 * replaced vauth_getpw() with inquery()
 *
 * Revision 2.1  2002-04-15 20:14:04+05:30  Cprogrammer
 * prevent authentication if NO_RELAY is set
 *
 * Revision 1.1  2002-03-18 22:26:23+05:30  Cprogrammer
 * Initial revision
 *
 */
#include "indimail.h"
#include <stdlib.h>
#include <string.h>
#define _XOPEN_SOURCE
#include <unistd.h>
#include <errno.h>

#ifndef lint
static char     sccsid[] = "$Id: vchkpass.c,v 2.33 2009-11-15 14:40:32+05:30 Cprogrammer Stab mbhangui $";
#endif
#ifdef AUTH_SIZE
#undef AUTH_SIZE
#define AUTH_SIZE 512
#endif

int             authlen = AUTH_SIZE;

int
main(int argc, char **argv)
{
	char           *tmpbuf, *login, *response, *challenge, *crypt_pass, *ptr, *cptr;
	char            user[AUTH_SIZE], domain[AUTH_SIZE], buf[MAX_BUFF];
	int             count, offset, norelay = 0, status;
	struct passwd  *pw;

	if (argc < 2)
		_exit(2);
	if (!(tmpbuf = calloc(1, (authlen + 1) * sizeof(char))))
	{
		printf("454-%s (#4.3.0)\r\n", strerror(errno));
		fflush(stdout);
		fprintf(stderr, "malloc-%d: %s\n", authlen + 1, strerror(errno));
		_exit(111);
	}
	for (offset = 0;;)
	{
		do
		{
			count = read(3, tmpbuf + offset, authlen + 1 - offset);
#ifdef ERESTART
		} while (count == -1 && (errno == EINTR || errno == ERESTART));
#else
		} while (count == -1 && errno == EINTR);
#endif
		if (count == -1)
		{
			printf("454-%s (#4.3.0)\r\n", strerror(errno));
			fflush(stdout);
			fprintf(stderr, "read: %s\n", strerror(errno));
			_exit(111);
		} else
		if (!count)
			break;
		offset += count;
		if (offset >= (authlen + 1))
			_exit(2);
	}
	count = 0;
	login = tmpbuf + count; /*- username */
	for (;tmpbuf[count] && count < offset;count++);
	if (count == offset || (count + 1) == offset)
		_exit(2);
	count++;
	challenge = tmpbuf + count; /*- challenge (or plain text) */
	for (;tmpbuf[count] && count < offset;count++);
	if (count == offset || (count + 1) == offset)
		_exit(2);
	response = tmpbuf + count + 1; /*- response */
	for (cptr = user, ptr = login;*ptr && *ptr != '@';*cptr++ = *ptr++);
	*cptr = 0;
	if (*ptr)
		ptr++;
	else
		getEnvConfigStr(&ptr, "DEFAULT_DOMAIN", DEFAULT_DOMAIN);
	for (cptr = domain;*ptr;*cptr++ = *ptr++);
	*cptr = 0;
#ifdef QUERY_CACHE
	if (!getenv("QUERY_CACHE"))
	{
#ifdef CLUSTERED_SITE
		if (vauthOpen_user(login))
#else
		if (vauth_open((char *) 0))
#endif
		{
			if (userNotFound)
				pipe_exec(argv, tmpbuf, offset);
			else
				fprintf(stderr, "vauth_open: %s\n", strerror(errno));
			printf("454-failed to connect to database (%s) (#4.3.0)\r\n", strerror(errno));
			fflush(stdout);
			_exit (111);
		}
		pw = vauth_getpw(user, domain);
		vclose();
	} else
		pw = inquery(PWD_QUERY, login, 0);
#else
#ifdef CLUSTERED_SITE
	if (vauthOpen_user(login))
#else
	if (vauth_open((char *) 0))
#endif
	{
		if (userNotFound)
			pipe_exec(argv, tmpbuf, offset);
		else
			fprintf(stderr, "vauth_open: %s\n", strerror(errno));
		printf("454-failed to connect to database (%s) (#4.3.0)\r\n", strerror(errno));
		fflush(stdout);
		_exit (111);
	}
	pw = vauth_getpw(user, domain);
	vclose();
#endif
	if (!pw)
	{
		if (userNotFound)
			pipe_exec(argv, tmpbuf, offset);
		else
			fprintf(stderr, "vchkpass: inquery: %s\n", strerror(errno));
		printf("454-%s (#4.3.0)\r\n", strerror(errno));
		fflush(stdout);
		_exit (111);
	} else
	if (pw->pw_gid & NO_SMTP)
	{
		printf("553-Sorry, this account cannot use SMTP (#5.7.1)\r\n");
		fflush(stdout);
		_exit (1);
	} else
	if (is_inactive && !getenv("ALLOW_INACTIVE"))
	{
		printf("553-Sorry, this account has expired (#5.7.1)\r\n");
		fflush(stdout);
		_exit (1);
	} else
	if (pw->pw_gid & NO_RELAY)
		norelay = 1;
	crypt_pass = pw->pw_passwd;
	if (getenv("DEBUG"))
	{
		fprintf(stderr, "%s: login [%s] challenge [%s] response [%s] pw_passwd [%s]\n", 
			argv[0], login, challenge, response, crypt_pass);
	}
	if (pw_comp((unsigned char *) login, (unsigned char *) crypt_pass,
		(unsigned char *) challenge, (unsigned char *) response))
	{
		pipe_exec(argv, tmpbuf, offset);
		printf("454-%s (#4.3.0)\r\n", strerror(errno));
		fflush(stdout);
		_exit (111);
	}
	status = 0;
	if ((ptr = (char *) getenv("POSTAUTH")) && !access(ptr, X_OK))
	{
		snprintf(buf, MAX_BUFF, "%s %s", ptr, login);
		status = runcmmd(buf, 0);
	}
	_exit(norelay ? 3 : status);
	/*- Not reached */
	return(0);
}

void
getversion_vchkpass_c()
{
	printf("%s\n", sccsid);
	printf("%s\n", sccsidh);
}
